### PR TITLE
OSD-25491,OSD-25492:New recording rules averaging CPU consumption for ROSA HCP & ROSA classic clusters

### PR DESCRIPTION
**Pull request by Nikokolas3270**
Those recording rules will be used for the (new) billing of ROSA

Closes [OSD-25491](https://issues.redhat.com//browse/OSD-25491) and [OSD-25492](https://issues.redhat.com//browse/OSD-25492)

### CODE REVIEW

**Review by simonpasquier — COMMENTED** [review]
Can we have unit tests for the new rules like we had in #518 and #506?

**Review by saswatamcode — APPROVED** [review]
LGTM

**Comment by Nikokolas3270** [comment]
Unit-tests added as requested by Simon

File: jsonnet/telemeter/rules.libsonnet
Comment (simonpasquier) [inline jsonnet/telemeter/rules.libsonnet:211]: (suggestion) "...vcpus:vcpu..." feels redundant. Also https://prometheus.io/docs/practices/rules/#naming
```suggestion
              record: 'hostedcluster:hypershift_cluster_vcpus:avg1h',
```

File: jsonnet/telemeter/rules.libsonnet
Comment (simonpasquier) [inline jsonnet/telemeter/rules.libsonnet:208]: HyperShift == ROSA HCP? Since we mention "ROSA HCP" below, it might be good to use the same terminology here.

File: jsonnet/telemeter/rules.libsonnet
Comment (simonpasquier) [inline jsonnet/telemeter/rules.libsonnet:217]: Maybe worth noting that we prefer the ROSA HCP metric if both exist for the same cluster _id?

**Review by simonpasquier — COMMENTED** [review]
/lgtm
/hold
I have some cosmetic remarks but nothing blocking (especially since I'm not using these metrics directly).

**Comment by simonpasquier** [comment]
/retitle OSD-25491,OSD-25492:New recording rules averaging CPU consumption for ROSA HCP & ROSA classic clusters

**Comment by simonpasquier** [comment]
@Nikokolas3270 feel free to unblock if you want (again my comments were nice-to-have).

File: jsonnet/telemeter/rules.libsonnet
Comment (Nikokolas3270) [reply jsonnet/telemeter/rules.libsonnet:213]: `scalar(count_over_time(vector(1)[1h:5m])` is not always equals to 12. It oscillate between 12 and 13.
Also, even it was always equal to 12, replacing it by `12` will feel like a hard code; not to mention that several other recording rules already use the same kind of construct in this file.

File: jsonnet/telemeter/rules.libsonnet
Comment (Nikokolas3270) [reply jsonnet/telemeter/rules.libsonnet:211]: Naming `hostedcluster:hypershift_cluster_vcpus:vcpu_hours` recording rule that way has been requested in [OSD-25491](https://issues.redhat.com/browse/OSD-25491). I agree that the name is not ideal (repeating the `vcpu` part...); that said it is consistent with the other recording rules in the file "averaging" CPU consumption during the last hour.
Also, naming this recording rule that way is quite useful to make sure that `rosa:cluster:vcpu_hours` recording rule, based on it, is not comparing apples and oranges.
I am not against changing this suffix, but if we do it or start to do it, lets make sure it is consistently changed everywhere.

File: jsonnet/telemeter/rules.libsonnet
Comment (Nikokolas3270) [reply jsonnet/telemeter/rules.libsonnet:208]: Well that's a tricky question.
Telemeter "hosts" metrics for all kind of clusters, including unmanaged clusters.
For this reason, the comment on top of `cluster:usage:workload:capacity_virtual_cpu_hours` definition is not saying that it concerns ROSA classic only... Because in fact this recording rule is defined and takes values for all kind of clusters, including ROSA classic but also OSD clusters or even OCP clusters installed on customer premises.
The "sister" `hostedcluster:hypershift_cluster_vcpus:vcpu_hours` recording rule suffers from the same fate: it is defined on cluster running the hypershift operator. This includes "our" management clusters hosting the ROSA HCP clusters but this could include as well any customer cluster running the hypershift operator.
So Hypershift != ROSA HCP.
With that in mind, this raises the concern that recording rule `rosa:cluster:vcpu_hours` is maybe poorly named. Indeed this rule is defined for:
- OSD clusters which are not ROSA
- OCP clusters installed and managed by the customers
- Clusters hosted by Hypershift but which are managed by the customer.

File: jsonnet/telemeter/rules.libsonnet
Comment (Nikokolas3270) [reply jsonnet/telemeter/rules.libsonnet:217]: Well, I think that the expression is pretty clear about that.
But I agree that this may bother people not familiar with Prometheus expression. I will add a comment in that direction.

Comment (simonpasquier) [reply jsonnet/telemeter/rules.libsonnet:217]: From experience, the more information we add here the more future ourselves will be grateful :)

File: jsonnet/telemeter/rules.libsonnet
Comment (simonpasquier) [reply jsonnet/telemeter/rules.libsonnet:213]: > `scalar(count_over_time(vector(1)[1h:5m])` is not always equals to 12. It oscillate between 12 and 13.

ok now I remember this glitch in Thanos. In theory, it should always return 12 which is why I suggested to hardcode but indeed it's not possible.

**Review by simonpasquier — COMMENTED** [review]
/hold cancel

My remaining remarks are only about code comments, no need to block the PR on them.

File: jsonnet/telemeter/rules.libsonnet
Comment (Nikokolas3270) [reply jsonnet/telemeter/rules.libsonnet:211]: Yeah, I have started a thread with the author of the JIA ticket.
We should eventually consider refactoring those rules from time to time and better name the rules ans the metrics.

File: jsonnet/telemeter/rules.libsonnet
Comment (Nikokolas3270) [reply jsonnet/telemeter/rules.libsonnet:210]: Corrective PR:
https://github.com/openshift/telemeter/pull/546

File: jsonnet/telemeter/rules.libsonnet
Comment (Nikokolas3270) [reply jsonnet/telemeter/rules.libsonnet:217]: Corrective PR:
https://github.com/openshift/telemeter/pull/546

File: jsonnet/telemeter/rules.libsonnet
Comment (kahowell) [reply jsonnet/telemeter/rules.libsonnet:208]: FWIW, swatch is a single consumer of this metric, and we can easily handle whatever name is most clear to others. So I'm inclined to name precisely based on what the metric is, rather than its intended use.

**Pull request merged**
